File: client/src/core/input/InputController.java
package core.input;

import core.base.NamedComponent;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.awt.*;
import java.awt.event.*;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;

import static core.input.InputEvent.InputEventType.*;

public class InputController implements NamedComponent {

	protected final Logger logger = LogManager.getLogger(getClass());

	// CONSTANTS
	public static final int MOUSE_LEFT = 0x0001_0000;
	public static final int MOUSE_CENTER = 0x0002_0000;
	public static final int MOUSE_RIGHT = 0x0003_0000;
	public static final int MOUSE_SCROLL = 0x0004_0000;

	// EVENT PROCESSING
	private List<InputEvent> queuedInputEvents;

	// CACHE
	private HashSet<Integer> holdKeyCodes;
	private Point currentMousePosition;
	private InputHandler currentlyFocused;

	public InputController() {
		queuedInputEvents = new ArrayList<>();
		this.holdKeyCodes = new HashSet<>();
		this.currentMousePosition = new Point(-1, -1);
		this.hoveringEnabled = true;
	}

	public List<InputEvent> popAllEvents() {
		List<InputEvent> result = queuedInputEvents;
		queuedInputEvents = new ArrayList<>();
		return result;
	}

	public InputHandler getCurrentlyFocused() {
		return currentlyFocused;
	}

	// ADAPTERS
	private KeyAdapter keyboard = new KeyAdapter() {
		@Override
		public void keyPressed(KeyEvent e) {
			int keycode = e.getKeyCode();
			if (keycode == 0) keycode = e.getExtendedKeyCode();
			Character c = e.getKeyChar();
			if (!Character.isDefined(c)) c = null;
			queuedInputEvents.add(new InputEvent(keycode, PRESS, 0, new HashSet<>(holdKeyCodes), currentMousePosition, c));

			holdKeyCodes.add(e.getKeyCode());
		}

		@Override
		public void keyReleased(KeyEvent e) {
			int keycode = e.getKeyCode();
			if (keycode == 0) keycode = e.getExtendedKeyCode();
			holdKeyCodes.remove(e.getKeyCode());
			queuedInputEvents.add(new InputEvent(keycode, RELEASE, 0, new HashSet<>(holdKeyCodes), currentMousePosition));
		}
	};

	public KeyAdapter getKeyboard() {
		return keyboard;
	}
	private MouseAdapter mouse = new MouseAdapter() {

		@Override
		public void mouseDragged(MouseEvent e) {
			this.mouseMoved(e);
		}

		@Override
		public void mouseMoved(MouseEvent e) {
			Point p = e.getPoint();
			if (p == null) return;
			currentMousePosition = p;
		}

		@Override
		public void mousePressed(MouseEvent e) {
			int code = btnToCode(e.getButton());
			queuedInputEvents.add(new InputEvent(code, PRESS, 0, holdKeyCodes, currentMousePosition));
			holdKeyCodes.add(code);
		}

		@Override
		public void mouseReleased(MouseEvent e) {
			int code = btnToCode(e.getButton());
			holdKeyCodes.remove(code);
			queuedInputEvents.add(new InputEvent(code, RELEASE, 0, holdKeyCodes, currentMousePosition));
		}

		@Override
		public void mouseWheelMoved(MouseWheelEvent e) {
			int code = MOUSE_SCROLL;
			queuedInputEvents.add(new InputEvent(code, CHANGE, e.getPreciseWheelRotation(), holdKeyCodes, currentMousePosition));
		}

		private int btnToCode(int button) {
			if (button == MouseEvent.BUTTON1) return MOUSE_LEFT;
			if (button == MouseEvent.BUTTON2) return MOUSE_CENTER;
			if (button == MouseEvent.BUTTON3) return MOUSE_RIGHT;
			logger.warn("Unknown mouse button: " + button);
			return MOUSE_CENTER;
		}

	};

	public MouseAdapter getMouse() {
		return mouse;
	}
	private boolean hoveringEnabled;

	/**
	 * @return mouse coordinate, DO NOT MODIFY THIS OBJECT!
	 */
	public Point getMousePosition() {
		if (!this.hoveringEnabled) {
			return new Point(-2, -2);
		}
		return this.currentMousePosition;
	}

	public void setFocus(InputHandler handler) {
		this.currentlyFocused = handler;
	}

	public void unfocus(InputHandler handler) {
		if (this.currentlyFocused == handler) this.currentlyFocused = null;
	}

	public void handleInput(InputEvent event) {
		if (this.currentlyFocused != null) {
			if (event.isPressActionAndConsume(InputAction.DESELECT)) {
				this.currentlyFocused.unfocus();
				this.currentlyFocused = null;
			} else {
				this.currentlyFocused.handleInput(event);
			}
		}
	}

	public void setHoverEnabled(boolean b) {
		this.hoveringEnabled = b;
	}

}

File: client/src/ch/judos/snakes/client/core/io/InputController.kt
package ch.judos.snakes.client.core.io

import ch.judos.snakes.client.core.base.NamedComponent
import ch.judos.snakes.client.core.io.InputEvent.InputEventType
import org.apache.logging.log4j.LogManager
import java.awt.Point
import java.awt.event.*
import java.util.*

class InputController : NamedComponent {
	protected val logger = LogManager.getLogger(javaClass)

	// EVENT PROCESSING
	private var queuedInputEvents: MutableList<InputEvent> = ArrayList()
	private var hoveringEnabled: Boolean = true

	// CACHE
	private val holdKeyCodes: HashSet<Int> = HashSet()
	private var currentMousePosition: Point = Point(-1, -1)
	var currentlyFocused: InputHandler? = null
		private set

	fun popAllEvents(): List<InputEvent> {
		val result: List<InputEvent> = queuedInputEvents
		queuedInputEvents = ArrayList()
		return result
	}

	// ADAPTERS
	val keyboard: KeyAdapter = object : KeyAdapter() {
		override fun keyPressed(e: KeyEvent) {
			var keycode = e.keyCode
			if (keycode == 0) keycode = e.extendedKeyCode
			var c: Char? = e.keyChar
			if (!Character.isDefined(c!!)) c = null
			queuedInputEvents.add(InputEvent(keycode, InputEventType.PRESS, 0.0, HashSet(holdKeyCodes), currentMousePosition, c))
			holdKeyCodes.add(e.keyCode)
		}

		override fun keyReleased(e: KeyEvent) {
			var keycode = e.keyCode
			if (keycode == 0) keycode = e.extendedKeyCode
			holdKeyCodes.remove(e.keyCode)
			queuedInputEvents.add(InputEvent(keycode, InputEventType.RELEASE, 0.0, HashSet(holdKeyCodes), currentMousePosition))
		}
	}
	val mouse: MouseAdapter = object : MouseAdapter() {
		override fun mouseDragged(e: MouseEvent) {
			mouseMoved(e)
		}

		override fun mouseMoved(e: MouseEvent) {
			val p = e.point ?: return
			currentMousePosition = p
		}

		override fun mousePressed(e: MouseEvent) {
			val code = btnToCode(e.button)
			queuedInputEvents.add(InputEvent(code, InputEventType.PRESS, 0.0, holdKeyCodes, currentMousePosition))
			holdKeyCodes.add(code)
		}

		override fun mouseReleased(e: MouseEvent) {
			val code = btnToCode(e.button)
			holdKeyCodes.remove(code)
			queuedInputEvents.add(InputEvent(code, InputEventType.RELEASE, 0.0, holdKeyCodes, currentMousePosition))
		}

		override fun mouseWheelMoved(e: MouseWheelEvent) {
			val code = MOUSE_SCROLL
			queuedInputEvents.add(InputEvent(code, InputEventType.CHANGE, e.preciseWheelRotation, holdKeyCodes, currentMousePosition))
		}

		private fun btnToCode(button: Int): Int {
			if (button == MouseEvent.BUTTON1) return MOUSE_LEFT
			if (button == MouseEvent.BUTTON2) return MOUSE_CENTER
			if (button == MouseEvent.BUTTON3) return MOUSE_RIGHT
			logger.warn("Unknown mouse button: $button")
			return MOUSE_CENTER
		}
	}

	/**
	 * @return mouse coordinate, DO NOT MODIFY THIS OBJECT!
	 */
	val mousePosition: Point
		get() = if (!hoveringEnabled) {
			Point(-2, -2)
		} else currentMousePosition

	fun setFocus(handler: InputHandler?) {
		currentlyFocused = handler
	}

	fun unfocus(handler: InputHandler) {
		if (currentlyFocused === handler) currentlyFocused = null
	}

	fun handleInput(event: InputEvent) {
		if (currentlyFocused != null) {
			if (event.isPressActionAndConsume(InputAction.DESELECT)) {
				currentlyFocused!!.unfocus()
				currentlyFocused = null
			} else {
				currentlyFocused!!.handleInput(event)
			}
		}
	}

	fun setHoverEnabled(b: Boolean) {
		hoveringEnabled = b
	}

	companion object {
		// CONSTANTS
		const val MOUSE_LEFT = 0x00010000
		const val MOUSE_CENTER = 0x00020000
		const val MOUSE_RIGHT = 0x00030000
		const val MOUSE_SCROLL = 0x00040000
	}
}